drivers/docker/stats.go

```diff
@@ -0,0 +1,176 @@
+package docker
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"runtime"
+	"time"
+
+	docker "github.com/fsouza/go-dockerclient"
+	cstructs "github.com/hashicorp/nomad/client/structs"
+	"github.com/hashicorp/nomad/helper/stats"
+	nstructs "github.com/hashicorp/nomad/nomad/structs"
+)
+
+const (
+	// statsCollectorBackoffBaseline is the baseline time for exponential
+	// backoff while calling the docker stats api.
+	statsCollectorBackoffBaseline = 5 * time.Second
+
+	// statsCollectorBackoffLimit is the limit of the exponential backoff for
+	// calling the docker stats api.
+	statsCollectorBackoffLimit = 2 * time.Minute
+)
+
+// Stats starts collecting stats from the docker daemon and sends them on the
+// returned channel.
+func (h *taskHandle) Stats(ctx context.Context, interval time.Duration) (<-chan *cstructs.TaskResourceUsage, error) {
+	select {
+	case <-h.doneCh:
+		return nil, nstructs.NewRecoverableError(fmt.Errorf("container stopped"), false)
+	default:
+	}
+	ch := make(chan *cstructs.TaskResourceUsage, 1)
+	go h.collectStats(ctx, ch, interval)
+	return ch, nil
+}
+
+// collectStats starts collecting resource usage stats of a docker container
+func (h *taskHandle) collectStats(ctx context.Context, ch chan *cstructs.TaskResourceUsage, interval time.Duration) {
+	defer close(ch)
+	// backoff and retry used if the docker stats API returns an error
+	var backoff time.Duration
+	var retry int
+	// loops until doneCh is closed
+	for {
+		if backoff > 0 {
+			select {
+			case <-time.After(backoff):
+			case <-ctx.Done():
+				return
+			case <-h.doneCh:
+				return
+			}
+		}
+		// make a channel for docker stats structs and start a collector to
+		// receive stats from docker and emit nomad stats
+		// statsCh will always be closed by docker client.
+		statsCh := make(chan *docker.Stats)
+		go dockerStatsCollector(ch, statsCh, interval)
+
+		statsOpts := docker.StatsOptions{
+			ID:      h.containerID,
+			Context: ctx,
+			Done:    h.doneCh,
+			Stats:   statsCh,
+			Stream:  true,
+		}
+
+		// Stats blocks until an error has occurred, or doneCh has been closed
+		if err := h.client.Stats(statsOpts); err != nil && err != io.ErrClosedPipe {
```

Comment:
ErrClosedPipe indicates the Read-side closed, but do we also need to ignore io.EOF for when the Write-side closes?

Reply:
We actually don't get an EOF. The docker api catches that and returns nil since thats the indicator that the container has stopped.